### PR TITLE
CompletionStage interoperability

### PR DESCRIPTION
**Pull request by jponge**
This is a proposal for a `Future` and `CompletionStage` interoperability.

### CODE REVIEW

**Comment by jponge** [comment]
Todo

[ ] Examples
[ ] Asciidoc documentation

File: src/main/java/io/vertx/core/Future.java [hunk 1]
```diff
@@ -479,6 +484,71 @@ default Context context() {
...
+        promise.complete(value);
+      }
+    });
+    return promise.future();
+  }
+
+  /**
+   * Bridges a {@link CompletionStage} object to a Vert.x future instance.
+   * <p>
+   * The Vert.x future handling methods will be called from a thread attached to {@code context}.
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/Future.java]: Minor, wording
```suggestion
   * The Vert.x future handling methods will be called on the provided {@code context}.
```

**Comment by jponge** [comment]
@tsegismont the PR is now ready for proper review

File: src/main/asciidoc/index.adoc [hunk 2]
```diff
@@ -361,6 +361,31 @@ run the given function, that returns a future. When this returned future complet
...
+. the first variant takes just a `CompletionStage` and calls the `Future` methods from the thread that resolves the `CompletionStage` instance, and
+. the second variant takes an extra {@link io.vertx.core.Context} parameter to call the `Future` methods on a Vert.x context.
+
+Here is an example of going from a `CompletionStage` to a Vert.x `Future` and dispatching on a context:
+
+[source,$lang]
+----
+{@link examples.CompletionStageInteropExamples#fromCS}
+----
+
```
Comment (tsegismont) [inline src/main/asciidoc/index.adoc]: Can you add an `IMPORTANT` note here to explain that in the verticle, using the second form is most often the way to go?

File: src/main/java/examples/CompletionStageInteropExamples.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        System.err.println("Could not resolve vertx.io");
+        err.printStackTrace();
+      } else {
+        System.out.println("vertx.io => " + ip);
+      }
+    });
+  }
+
+  public void fromCS(Vertx vertx, CompletionStage<String> completionStage) {
+    Future.from(completionStage, vertx.getOrCreateContext())
```
Comment (tsegismont) [inline src/main/java/examples/CompletionStageInteropExamples.java]: This example doesn't give the user a hint about when to do this.

How about using the CS result in a future `flatMap` operation?

File: src/test/java/io/vertx/core/FutureTest.java [hunk 4]
```diff
@@ -1424,4 +1424,133 @@ public void testFailureNotification() {
     promise.fail(failure);
     await();
   }
+
+  @Test
+  public void testToCompletionStageTrampolining() {
```
Comment (tsegismont) [inline src/test/java/io/vertx/core/FutureTest.java]: Not mandatory but we could save some duplication by refactoring these tests

File: src/main/asciidoc/index.adoc (hunk 2 above)
Comment (jponge) [reply src/main/asciidoc/index.adoc]: Added

File: src/main/java/examples/CompletionStageInteropExamples.java (hunk 3 above)
Comment (jponge) [reply src/main/java/examples/CompletionStageInteropExamples.java]: I've modified the example, you'll let me know how it goes

File: src/test/java/io/vertx/core/FutureTest.java (hunk 4 above)
Comment (jponge) [reply src/test/java/io/vertx/core/FutureTest.java]: I think they are readable as-is

Comment (tsegismont) [reply src/test/java/io/vertx/core/FutureTest.java]: ok

**Comment by jponge** [comment]
Ping @vietj

**Comment by tsegismont** [comment]
@vietj could you please take a look? This would be helpful in vert-x3/vertx-cassandra-client#45

**Comment by jponge** [comment]
Also in the book :-)

File: src/test/java/io/vertx/core/FutureTest.java [hunk 5]
```diff
@@ -1424,4 +1424,133 @@ public void testFailureNotification() {
...
+      assertTrue(err instanceof RuntimeException);
+      assertEquals("Woops", err.getMessage());
+      assertNotSame(mainThread, Thread.currentThread());
+      complete();
+    });
+
+    disableThreadChecks();
+    new Thread(() -> willSucceed.complete("Yo")).start();
+    new Thread(() -> willFail.fail(new RuntimeException("Woops"))).start();
+    await(5, SECONDS);
```
Comment (vietj) [inline src/test/java/io/vertx/core/FutureTest.java]: you don't need to set the number of seconds here (also 5 seconds might timeout in slow CIs on a GC), simply `await()` will do the job, there is default timeout set of 2 minutes

File: src/main/java/io/vertx/core/Future.java [hunk 6]
```diff
@@ -479,6 +481,71 @@ default Context context() {
...
+   * Bridges a {@link CompletionStage} object to a Vert.x future instance.
+   * <p>
+   * The Vert.x future handling methods will be called from the thread that completes {@code completionStage}.
+   *
+   * @param completionStage a completion stage
+   * @param <T>             the result type
+   * @return a Vert.x future that resolves when {@code completionStage} resolves
+   */
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  static <T> Future<T> from(CompletionStage<T> completionStage) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/Future.java]: can you rename to `fromCompletionStage` ?

Comment (tsegismont) [reply src/main/java/io/vertx/core/Future.java]: @vietj I had the same comment at first but then I agreed with @jponge that a compact form is more readable.

Comment (vietj) [reply src/main/java/io/vertx/core/Future.java]: in a previous conversation with @jponge we agreed on `fromCompletionStage` (see https://github.com/vert-x3/wiki/wiki/RFC:-Future-API#open-questions--todo) when we drafted this API. I think the main reason is that it was symetrical to `toCompletionStage` and be more readable for `CompletionStage` users

Comment (jponge) [reply src/main/java/io/vertx/core/Future.java]: When we pair-programmed with @tsegismont I thought `from` was nice. But I can rename.

File: src/test/java/io/vertx/core/FutureTest.java (hunk 5 above)
Comment (jponge) [reply src/test/java/io/vertx/core/FutureTest.java]: Ok

**Comment by jponge** [comment]
@tsegismont @vietj updated

**Comment by tsegismont** [comment]
@jponge @vietj fine with me. This feature is more important than the name.

**Comment by jponge** [comment]
@vietj can I merge now?

**Comment by vietj** [comment]
+1

**Comment by jponge** [comment]
Let's go :-)

**Pull request merged**